quark/subproject.py

```diff
@@ -705,6 +713,287 @@ def set_local_ignores(self, subprojects_dir, modules):
+
+            parts = url.path.split("/")
+
+            self.parsed_project_name = "/".join(parts[0 : parts.index("artifacts")]).strip("/")
+            self.parsed_artifact_name = parts[-1]
+            self.parsed_job = fragments.get("job")
+            self.parsed_ref = fragments.get("ref")
+
+            artifact_path = "/".join(parts[parts.index("artifacts") + 1 :])
+
+            if "ref" in fragments and not self.gitlab_token_is_pat:
+                # NOTE: We have "ref" and "job" but we don't have a Personal Access Token (PAT). We
+                # can't resolve the job ID using the API and a CI job token can't be used for this.
+                #
+                # TODO: Find a way to resolve the job ID without requiring private tokens so that we
+                # can use a single endpoint (and we can always store the job ID inside the stamp
+                # file).
+                self.parsed_endpoint_url = "/projects/%s/jobs/artifacts/%s/raw/%s?job=%s" % (
+                    urllib.parse.quote_plus(self.parsed_project_name),
+                    urllib.parse.quote_plus(fragments["ref"]),
+                    artifact_path,
+                    urllib.parse.quote_plus(fragments["job"]),
+                )
+            else:
+                if "ref" in fragments and self.gitlab_token_is_pat:
+                    print_msg("resolving job id for " + self.parsed_artifact_name, self._comment())
+                    self.stamp["job_id"] = self._resolve_job_id(
+                        self.parsed_project_name,
+                        fragments["ref"],
+                        fragments["job"],
+                    )
+                else:
+                    self.stamp["job_id"] = str(fragments["job"])
+
+                self.parsed_endpoint_url = "/projects/%s/jobs/%s/artifacts/%s" % (
+                    urllib.parse.quote_plus(self.parsed_project_name),
+                    self.stamp["job_id"],
+                    artifact_path,
+                )
+        elif url.scheme == "gitlab+package":
+            project_name, package_name, package_version, package_file_name = url.path.rsplit("/", 3)
+
+            self.parsed_project_name = project_name.strip("/")
+            self.parsed_artifact_name = package_file_name
+            self.parsed_pkg = "%s/%s/%s" % (package_name, package_version, package_file_name)
+            self.parsed_endpoint_url = "/projects/%s/packages/generic/%s" % (
+                urllib.parse.quote_plus(self.parsed_project_name),
+                self.parsed_pkg,
+            )
+        else:
+            raise QuarkError("Unsupported URL: " + url)
+
+    def _resolve_job_id(self, project: str, ref: str, job_name: str) -> str:
+        # NOTE: This method requires an API token. A CI job token isn't sufficient. As such, this
+        # should be called only when freezing (usually done by a developer with a private token).
```

Comment:
> usually done by a developer with a private token

Unfortunately on pretty much all CI jobs we do have a final "freeze + archive freeze file as artifact" step, used to remember what was actually downloaded in that specific build (that's important for builds that run on branches without a committed freeze.quark, such as most of our `master` or release branches).